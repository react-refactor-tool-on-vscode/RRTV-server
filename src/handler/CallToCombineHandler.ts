import {
    CodeAction,
    CodeActionParams,
    Command,
    Range,
    TextDocumentChangeEvent,
    TextDocumentItem,
    TextDocuments,
    TextEdit,
    CodeActionKind,
    ExecuteCommandParams,
    TextDocumentEdit,
    ExecuteCommandRequest,
    Position,
    WorkspaceChange,
} from "vscode-languageserver/node";
import { BaseHandler, ContinuousOutputHandler } from "../interface/Handler";
import { connection, documents } from "../server";
import { DocumentUri } from "vscode-languageserver-textdocument";
import { hasJSXCode, addTabStop } from "../helper/attrInsertor";
import { transformer } from "../helper/callToCombine";

export class CallToCombineHandler extends ContinuousOutputHandler<
    (CodeAction | Command)[],
    CodeActionParams
> {
    protected concreteHandle(
        prevOutput: (CodeAction | Command)[],
        request: CodeActionParams
    ): (CodeAction | Command)[] {
        try {
            const range = request.range;
            const uri = request.textDocument.uri;
            const document = documents.get(uri);
            const result = transformer(
                document.getText(),
                document.offsetAt(range.start)
            );
            console.log("result::: ", result);
            if (!result.check) {
                return prevOutput;
            }
            const change = new WorkspaceChange();
            const a = change.getTextEditChange(uri);
            a.replace(
                Range.create(Position.create(0, 0), Position.create(1000, 300)),
                result.newText
            );
            const codeAction = CodeAction.create(
                "Component Call to Component Combine",
                change.edit,
                CodeActionKind.RefactorRewrite
            );
            return [...(prevOutput ?? []), codeAction];
        } catch (error) {
            console.log(JSON.stringify(error));
        }
        return [...(prevOutput ?? [])];
    }
}
