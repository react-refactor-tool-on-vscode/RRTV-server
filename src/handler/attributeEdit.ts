import {
    CodeAction,
    CodeActionParams,
    Command,
    Range,
    TextDocumentChangeEvent,
    TextDocumentItem,
    TextDocuments,
    TextEdit,
    CodeActionKind,
    ExecuteCommandParams,
    TextDocumentEdit,
    ExecuteCommandRequest,
} from "vscode-languageserver/node";
import {BaseHandler, ContinuousOutputHandler} from '../interface/Handler'
import { connection, documents } from '../server'

class AttrEditHandler extends ContinuousOutputHandler<
    (CodeAction | Command)[],
    CodeActionParams
> {
    protected concreteHandle(prevOutput: (CodeAction | Command)[], request: CodeActionParams): (CodeAction | Command)[] {
        const codeAction = generateCodeAction(request);
        if(codeAction === undefined) return [...prevOutput];
        else return [...prevOutput, codeAction];
    }
}

function generateCodeAction(request:CodeActionParams): CodeAction | undefined {
    const text = documents.get(request.textDocument.uri).getText(request.range);
    if(!hasJSXCode(text)) return;
    const options = [
        "same key same value",
        "same key different value",
        "different key same value",
        "different key different value"
    ]
    const codeAction:CodeAction = {
        title: 'Add attributes to labels in batches',
        kind: CodeActionKind.Refactor,
        data: request.textDocument.uri,
        command: Command.create(
            'provide attribute',
            'provide attribute',
            {
                document: request.textDocument.uri,
                range: request.range,
            },
            options,
        )
    }
    return codeAction;
}

function addTabStop(text:string):string | undefined {
    if(!hasJSXCode(text)) return;
    const regex = /<(\w+)((?:\s+\w+(?:\s*=\s*(?:".*?"|'.*?'|[\^'">\s]+))?)+\s*|\s*)>/g;
    const modifiedCode = text.replace(regex, (match, tagName, attributes) => {
        const attributeString = ' $1=$2';
        return `<${tagName}${attributes}${attributeString}>`;
    });
    return modifiedCode;
}

function hasJSXCode(code: string): boolean {
    const regex = /<([a-zA-Z][\w.-]*)\b/g;  // 匹配以大写字母开头的标签名称
    return regex.test(code);
}

async function generateSnippet(params:ExecuteCommandParams) {
    if(!params.arguments) return;
    const uri = params.arguments[0].document;
    const range = params.arguments[0].range;
    const document = documents.get(uri);
    const text = document.getText(range);
    const modifiedCode = addTabStop(text);
    if(!modifiedCode) return;
    const documentChanges = [
		TextDocumentEdit.create({
			uri:uri,
			version:document.version
		},
			[TextEdit.del(range)]
		)
	];
    await connection.workspace.applyEdit({
		documentChanges: documentChanges
	});
    await connection.sendRequest(ExecuteCommandRequest.method, {
        command: "run snippet",
		arguments:[range.start, modifiedCode]
    })
}

export {AttrEditHandler, addTabStop, generateSnippet};
