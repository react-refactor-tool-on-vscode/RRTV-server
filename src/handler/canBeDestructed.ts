import traverse, { NodePath, TraverseOptions } from "@babel/traverse";
import * as t from "@babel/types";
import { Range } from "vscode-languageserver";
import { IsRangeInLoc } from "../helper/RangeLoc";
import { isHeadIdOfMemberExpr } from "../helper/IsHeadIdOfMemberExpr";

export function canBeDestructed(
    range: Range,
    ast: t.Node
): { beDestructed: boolean; identifierName: string; } {
    let isParamSelected: boolean = false;
    let identifierName: string;
    let hasTrailing: boolean = false;

    const findIdentifier: TraverseOptions<t.Node> = {
        Identifier(path) {
            if (IsRangeInLoc(range, path.node.loc)) {
                if (t.isObjectProperty(path.parent)) {
                    if (t.isIdentifier(path.parent.value) &&
                        path.node.name ==
                        (path.parent.value as t.Identifier).name) {
                        isParamSelected = true;
                        identifierName = path.node.name;
                    }
                } else if (t.isArrayPattern(path.parent)) {
                    isParamSelected = true;
                    identifierName = path.node.name;
                }
            }
        },
    };

    const checkHasTrailing: TraverseOptions<t.Node> = {
        Identifier(path) {
            if (isHeadIdOfMemberExpr(path)) {
                hasTrailing = true;
            }
        },
    };

    traverse(ast, {
        enter(path) {
            if (t.isFunction(path.node) && IsRangeInLoc(range, path.node.loc)) {
                let index: number;
                path.node.params.forEach((param, id) => {
                    if (IsRangeInLoc(range, param.loc)) {
                        index = id;
                    }
                });

                if (index == undefined) return;

                const paramChildPath = path.get(
                    `params.${index}`
                ) as NodePath<t.Node>;

                if (t.isIdentifier(paramChildPath.node)) {
                    isParamSelected = true;
                    identifierName = paramChildPath.node.name;
                } else {
                    paramChildPath.traverse(findIdentifier);
                }

                if (!isParamSelected) return; // Ensure that the range is on a param.


                // Check if it's references has trailing ones.
                const bodyPath = path.get("body") as NodePath<t.Node>;
                bodyPath.traverse(checkHasTrailing);
            }
        },
    });
    return {
        beDestructed: hasTrailing,
        identifierName: identifierName,
    };
}