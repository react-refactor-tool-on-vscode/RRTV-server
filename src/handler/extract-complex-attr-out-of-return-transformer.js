module.exports = (file, api, options) => {
    const j = api.jscodeshift;
    const root = j(file.source);
    let count = 0
    root.find(j.ArrowFunctionExpression)
        .forEach((path) => { handler(j, path, count++) })
    root.find(j.FunctionExpression)
        .forEach((path) => { handler(j, path, count++) })
    return root.toSource();
}

function handler(j, path, count) {
    const jsxAttributes = j(path).closest(j.JSXAttribute)
    if (!jsxAttributes.length) return
    const functionName = "NewFunction" + count
    const newFunction = j.functionDeclaration(
        j.identifier("NewFunction" + count),
        path.node.params,
        j.blockStatement(path.node.body.body)
    );
    jsxAttributes.closest(j.ReturnStatement)
        .insertBefore(newFunction)
    const newFunctionCall = j.callExpression(j.identifier(functionName), path.node.params);
    const newArrowFunctionExpression = j.arrowFunctionExpression(path.node.params, newFunctionCall)
    if (path.node.params.length === 0)
        j(path).replaceWith(j.identifier(functionName)) //ok
    else
        j(path).replaceWith(newArrowFunctionExpression)

}