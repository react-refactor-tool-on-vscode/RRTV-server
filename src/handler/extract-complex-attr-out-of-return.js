"use strict";
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
exports.__esModule = true;
var code = "const renderButton = (text, onClick) => {\n    return (\n        <>\n            <button onClick={() => {\n                console.log('Button clicked!');\n            }}>{text}</button>\n            <button onClick={function () {\n                console.log('Button clicked!')\n            }}></button>\n            <button onClick={newFunction}></button>\n            <button onClick={function (e) {\n                console.log('Button clicked!', e)\n            }}></button>\n            <button onClick={e => NewFunction(e)}></button>\n            <button onClick={function (e, ee) {\n                console.log('Button clicked!', ee, e)\n            }}></button>\n            <button onClick={(e) => {\n                console.log('Button clicked!', e);\n            }}>{text}</button>\n            <button onClick={function(e) {\n                return NewFunction00(e)\n            }}></button>\n        </>\n    );\n}";
var index = 120;
var jscodeshift = require("jscodeshift");
var transform = function (file, api, options) {
    var j = api.jscodeshift;
    var root = j(file.source);
    var count = 0;
    root.find(j.ArrowFunctionExpression)
        .filter(function (path) { return new handleType().arrowFunctionExpression(path) && handleIndex(index, path); })
        .forEach(function (path) { handler(j, path, count++); });
    root.find(j.FunctionExpression)
        .filter(function (path) { return new handleType().functinoExpression(path, j) && handleIndex(index, path); })
        .forEach(function (path) { handler(j, path, count++); });
    return root.toSource();
};
function handler(j, path, count) {
    var jsxAttributes = j(path).closest(j.JSXAttribute);
    if (!jsxAttributes.length)
        return;
    var functionName = "NewFunction" + count;
    var newFunction = j.functionDeclaration(j.identifier("NewFunction" + count), path.node.params, j.blockStatement(path.node.body.body));
    jsxAttributes.closest(j.ReturnStatement)
        .insertBefore(newFunction);
    var newFunctionCall = j.callExpression(j.identifier(functionName), path.node.params);
    var newArrowFunctionExpression = j.arrowFunctionExpression(path.node.params, newFunctionCall);
    if (path.node.params.length === 0)
        j(path).replaceWith(j.identifier(functionName)); //ok
    else
        j(path).replaceWith(newArrowFunctionExpression);
}
function handleIndex(index, path) {
    seePosition(path, index);
    if (path.node.start <= index && index <= path.node.end) {
        return true;
    }
    return false;
}
function seePosition(path) {
    var rest = [];
    for (var _i = 1; _i < arguments.length; _i++) {
        rest[_i - 1] = arguments[_i];
    }
    console.log.apply(console, __spreadArray([path.node.start, path.node.end], rest, false));
}
var handleType = /** @class */ (function () {
    function handleType() {
    }
    handleType.prototype.arrowFunctionExpression = function (path) {
        var rest = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            rest[_i - 1] = arguments[_i];
        }
        /// 不识别 <button onClick={e => NewFunction(e)}></button>
        var body = path.node.body;
        return (body.type === "JSXElement" ||
            body.type === "FunctionExpression" ||
            body.type === "ArrowFunctionExpression" ||
            body.type === "BlockStatement");
    };
    handleType.prototype.functinoExpression = function (path) {
        var rest = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            rest[_i - 1] = arguments[_i];
        }
        /// 不识别 <button onClick={function(e) { return NewFunction(e)}}></button>
        var body = path.node.body;
        /// j 放在 rest[0]
        var check = false;
        if (rest.length) {
            console.log(rest[0](path).find(rest[0].ReturnStatement).length ? rest[0](path).find(rest[0].ReturnStatement).get(0).node.argument.type : null);
            check = rest[0](path).find(rest[0].ReturnStatement, {
                argument: {
                    type: "CallExpression"
                }
            }).length > 0;
            console.log(check);
        }
        return !check;
    };
    return handleType;
}());
var output = transform({
    source: code,
    path: ""
}, {
    jscodeshift: jscodeshift,
    j: jscodeshift,
    stats: function () { },
    report: function (msg) {
    }
}, {});
console.log(output);
