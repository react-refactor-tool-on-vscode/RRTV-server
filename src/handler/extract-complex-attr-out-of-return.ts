const code = `const renderButton = (text, onClick) => {
    return (
        <>
            <button onClick={() => {
                console.log('Button clicked!');
            }}>{text}</button>
            <button onClick={function () {
                console.log('Button clicked!')
            }}></button>
            <button onClick={newFunction}></button>
            <button onClick={function (e) {
                console.log('Button clicked!', e)
            }}></button>
            <button onClick={e => NewFunction(e)}></button>
            <button onClick={function (e, ee) {
                console.log('Button clicked!', ee, e)
            }}></button>
            <button onClick={(e) => {
                console.log('Button clicked!', e);
            }}>{text}</button>
        </>
    );
}`

import * as jscodeshift from 'jscodeshift';

const transform = (file: jscodeshift.FileInfo, api: jscodeshift.API, options: jscodeshift.Options) => {
    const j = api.jscodeshift;
    const root = j(file.source);
    let count = 0
    root.find(j.ArrowFunctionExpression)
        .filter(path => new handleType().arrowFunctionExpression(path) && handleIndex(34, path))
        .forEach((path) => { handler(j, path, count++) })
    root.find(j.FunctionExpression)
        .forEach((path) => { handler(j, path, count++) })
    return root.toSource();
}

function handler(j: jscodeshift.JSCodeshift, path: any, count: number) {
    const jsxAttributes = j(path).closest(j.JSXAttribute)
    if (!jsxAttributes.length) return
    const functionName = "NewFunction" + count
    const newFunction = j.functionDeclaration(
        j.identifier("NewFunction" + count),
        path.node.params,
        j.blockStatement(path.node.body.body)
    );
    jsxAttributes.closest(j.ReturnStatement)
        .insertBefore(newFunction)
    const newFunctionCall = j.callExpression(j.identifier(functionName), path.node.params);
    const newArrowFunctionExpression = j.arrowFunctionExpression(path.node.params, newFunctionCall)
    if (path.node.params.length === 0)
        j(path).replaceWith(j.identifier(functionName)) //ok
    else
        j(path).replaceWith(newArrowFunctionExpression)

}

function handleIndex(index: number, path: any): boolean {
    console.log(path.node.start, path.node.end, index)
    return false
}

class handleType {
    constructor() { }
    arrowFunctionExpression(path: any) {
        const body = path.node.body;
        return (
            body.type === "JSXElement" ||
            body.type === "FunctionExpression" ||
            body.type === "ArrowFunctionExpression" ||
            body.type === "BlockStatement"
        );
    }
}

const output = transform(
    {
        source: code,
        path: ""
    },
    {
        jscodeshift,
        j: jscodeshift,
        stats: () => { },
        report(msg) {

        },
    },
    {}
)

console.log(output)

