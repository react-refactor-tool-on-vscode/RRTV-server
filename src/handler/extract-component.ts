import * as node from "vscode-languageserver/node";

import { TextDocument } from "vscode-languageserver-textdocument";

import { BaseHandler } from "../interface/Handler";

import { textToAst, dataGenerator, modifier } from './utils-extract-component';

export class ExtractComponentHandler extends BaseHandler<node.CodeAction[], node.CodeActionParams> {
    handle(prevOutput: node.CodeAction[], request: node.CodeActionParams): node.CodeAction[] {
        const documents: node.TextDocuments<TextDocument> = new node.TextDocuments(TextDocument)
        const document = documents.get(request.textDocument.uri)
        const text = document.getText(request.range)
        const ast = textToAst(text)
        const { valid, specified } = dataGenerator(ast)
        if (!valid) {
            return [...prevOutput]
        }

        const codeAction = node.CodeAction.create(
            "Extract Component",
            node.Command.create(
                "Extract Component",
                "rrtv.extractComponent",
                [{ ast, specified }] //{newText, range} = modifier(ast, specified)
            ),
            node.CodeActionKind.RefactorExtract
        );

        const {newText, _range} = modifier(ast, specified)
        const change: node.WorkspaceChange = new node.WorkspaceChange()
        const a = change.getTextEditChange(document)
        a.replace(node.Range.create(_range.start, _range.end), newText)

        codeAction.edit = change.edit
        return [...prevOutput, codeAction]
    }
}