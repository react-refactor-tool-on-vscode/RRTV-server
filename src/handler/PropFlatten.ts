import { BaseHandler, ContinuousOutputHandler } from "../interface/Handler";
import { parse } from "@babel/parser";
import {
    CodeAction,
    CodeActionKind,
    CodeActionParams,
    Command,
    ExecuteCommandParams,
    Range,
    TextDocumentChangeEvent,
    TextDocumentEdit,
    TextDocumentItem,
    TextDocuments,
    TextEdit,
} from "vscode-languageserver";
import { TextDocument } from "vscode-languageserver-textdocument";
import traverse from "@babel/traverse";

import { connection, documents } from "../server";
import parseToAst from "../helper/ParseToAst";
import { getTrailingIdsFromIdentifier } from "../helper/getTrailingIdsFromIdentifier";
import { canBeDestructed } from "../helper/canBeDestructed";
import { SourceLocation } from "@babel/types";
import { IsRangeInLoc } from "../helper/RangeLoc";
import { idsToTextEdit } from "../helper/idsToTextEdit";
import { locToRange } from "../helper/locToRange";

const commandName = "rrtv.propFlatten";

class PropFlattenCodeActionHandler extends ContinuousOutputHandler<
    (CodeAction | Command)[],
    CodeActionParams
> {
    protected concreteHandle(
        prevOutput: (CodeAction | Command)[],
        request: CodeActionParams
    ): (CodeAction | Command)[] {
        const codeAction = generateCodeAction(request);
        if (codeAction) {
            return [...(prevOutput ?? []), codeAction];
        } else return prevOutput;
    }
}

class PropFlattenExecuteCommandHandler extends BaseHandler<
    void,
    ExecuteCommandParams
> {
    handle(prevOutput: void, request: ExecuteCommandParams): void {
        if (request.command == commandName) {
            // Get input.
            const textDocument = documents.get(request.arguments[0]);
            const range: Range = request.arguments[1];
            const identifierName = request.arguments[2];
            const isDirectParam = request.arguments[3];
            // Send edits.
            connection.workspace.applyEdit({
                documentChanges: [
                    TextDocumentEdit.create(
                        {
                            uri: textDocument.uri,
                            version: textDocument.version,
                        },
                        generateTextEdit(textDocument, range, identifierName, isDirectParam)
                    ),
                ],
            });
            return;
        } else this.nextHandler.handle(null, request);
    }
}

function generateCodeAction(request: CodeActionParams): CodeAction | undefined {
    const range = request.range;
    const text = documents.get(request.textDocument.uri).getText();

    const ast = parseToAst(text);

    const canBeDestructedResult = canBeDestructed(range, ast);
    if (!canBeDestructedResult.beDestructed) return undefined;
    else
        return CodeAction.create(
            "Prop Flatten",
            Command.create(
                "Prop Flatten",
                commandName,
                request.textDocument.uri,
                request.range,
                canBeDestructedResult.identifierName,
                canBeDestructedResult.isDirectParam
            ),
            CodeActionKind.RefactorRewrite
        );
}

function generateTextEdit(
    document: TextDocument,
    range: Range,
    identifierName: string,
    isDirectParam: boolean
): TextEdit[] {
    const code = document.getText();
    const ast = parseToAst(code);
    let paramLoc: SourceLocation;
    traverse(ast, {
        Identifier(path) {
            if (IsRangeInLoc(range, path.node.loc)) {
                paramLoc = path.node.loc;
                return;
            }
        },
    });
    const trailingIds = getTrailingIdsFromIdentifier(
        ast,
        range,
        identifierName
    );

    return idsToTextEdit(
        document,
        Range.create(
            paramLoc.start.line - 1,
            paramLoc.start.column,
            paramLoc.end.line - 1,
            paramLoc.end.column
        ),
        trailingIds.ids,
        trailingIds.memberExprLocs.map((loc) => locToRange(loc)),
        isDirectParam
    );
}

export { PropFlattenCodeActionHandler, PropFlattenExecuteCommandHandler };
