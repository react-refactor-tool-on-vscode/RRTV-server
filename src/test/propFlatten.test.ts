import { parse } from "@babel/parser";
import * as t from "@babel/types";
import traverse, { NodePath } from "@babel/traverse";
import generate from "@babel/generator";
import { canBeDestructed } from "../handler/PropFlatten";
import { Range } from "vscode-languageserver";

const text1 = `function TodoItem({item: {source, text}}) {
    return (
        <>
        <img src={source} />
        <span>{text}</span>
        </>
    )
}`;

const ast1 = parse(text1, {
    sourceType: "module",
    errorRecovery: true,
    plugins: ["jsx"],
});

const text2 = `function TodoItem(item) {
    return (
        <>
        <img src={item.source} />
        <span>{item.text}</span>
        </>
    )
}`;

const ast2 = parse(text2, {
    sourceType: "module",
    errorRecovery: true,
    plugins: ["jsx"],
});

it("find all jsx identifier", () => {
    const jsxIdentifiers: any[] = [];

    traverse(ast1, {
        JSXIdentifier(path) {
            jsxIdentifiers.push(path.node.name);
        },
    });

    expect(jsxIdentifiers).toMatchSnapshot();
});

it("find all identifier", () => {
    const identifiers: any[] = [];

    traverse(ast1, {
        Identifier(path) {
            identifiers.push(path.node.name);
        },
    });

    expect(identifiers).toMatchSnapshot();
});

it("get params type", () => {
    let result: string;

    traverse(ast1, {
        enter(path) {
            if (t.isFunction(path.node)) {
                result = typeof path.get("params");
            }
        },
    });

    expect(result).toMatchSnapshot();
});

it("can get first param", () => {
    let result: boolean;

    traverse(ast1, {
        enter(path) {
            if (t.isFunction(path.node)) {
                result = typeof path.get("params.0") != null;

                const params = path.node.params;
                const first_param = path.get("params.0");
                const a = "asd";
            }
        },
    });

    expect(result).toMatchSnapshot();
});

it("functionTypeParam test", () => {
    const ast = t.functionTypeParam(
        t.identifier("id"),
        t.stringTypeAnnotation()
    );

    const code = generate(ast).code;

    expect(code).toMatchSnapshot();
});

it("path between parent and array(container)", () => {
    let result: boolean = false;
    const code = "function Test ([a,b]) {return a;}";
    const ast = parse(code);
    traverse(ast, {
        Identifier(path) {
            if (result == false) result = t.isArrayPattern(path.parent);
        },
    });

    expect(result).toMatchSnapshot();
});

it("traverse function body", () => {
    const identifiersInBody: any[] = [];

    traverse(ast1, {
        FunctionDeclaration(path) {
            path.get("body").traverse({
                Identifier(path) {
                    identifiersInBody.push(path.node.name);
                },
            });
        },
    });

    expect(identifiersInBody).toMatchSnapshot();
});

test("check param can be destructed", () => {
    expect(canBeDestructed(Range.create(0, 19, 0, 19), ast2)).toEqual({
        beDestructed: true,
        identifierName: "item",
    });

    expect(canBeDestructed(Range.create(0, 0, 0, 0), ast2)).toEqual({
        beDestructed: false,
        identifierName: null,
    });

    expect(canBeDestructed(Range.create(0, 18, 0, 21), ast2)).toEqual({
        beDestructed: true,
        identifierName: "item",
    });
});
