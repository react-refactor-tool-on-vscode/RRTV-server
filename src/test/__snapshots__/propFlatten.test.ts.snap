// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can get first param 1`] = `true`;

exports[`find all identifier 1`] = `
[
  "TodoItem",
  "item",
  "source",
  "source",
  "text",
  "text",
  "source",
  "text",
]
`;

exports[`find all jsx identifier 1`] = `
[
  "img",
  "src",
  "span",
  "span",
]
`;

exports[`functionTypeParam test 1`] = `"id: string"`;

exports[`get params type 1`] = `"object"`;

exports[`path between parent and array(container) 1`] = `true`;

exports[`traverse function body 1`] = `
[
  "source",
  "text",
]
`;
